// react-map-gl-draw/dist-es6/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EDIT_TYPE = exports.ELEMENT_TYPE = exports.GUIDE_TYPE = exports.RENDER_STATE = exports.RENDER_TYPE = exports.GEOJSON_TYPE = exports.DRAWING_MODE = exports.MODES = void 0;
var MODES = {
  READ_ONLY: 'READ_ONLY',
  SELECT: 'SELECT',
  EDITING: 'EDITING',
  DRAW_POINT: 'DRAW_POINT',
  DRAW_PATH: 'DRAW_PATH',
  DRAW_POLYGON: 'DRAW_POLYGON',
  DRAW_RECTANGLE: 'DRAW_RECTANGLE'
};
exports.MODES = MODES;
var DRAWING_MODE = [MODES.DRAW_POINT, MODES.DRAW_PATH, MODES.DRAW_POLYGON, MODES.DRAW_RECTANGLE];
exports.DRAWING_MODE = DRAWING_MODE;
var GEOJSON_TYPE = {
  POINT: 'Point',
  LINE_STRING: 'LineString',
  POLYGON: 'Polygon'
};
exports.GEOJSON_TYPE = GEOJSON_TYPE;
var RENDER_TYPE = {
  POINT: 'Point',
  LINE_STRING: 'LineString',
  POLYGON: 'Polygon',
  RECTANGLE: 'Rectangle'
};
exports.RENDER_TYPE = RENDER_TYPE;
var RENDER_STATE = {
  INACTIVE: 'INACTIVE',
  UNCOMMITTED: 'UNCOMMITTED',
  SELECTED: 'SELECTED',
  HOVERED: 'HOVERED',
  CLOSING: 'CLOSING'
};
exports.RENDER_STATE = RENDER_STATE;
var GUIDE_TYPE = {
  TENTATIVE: 'tentative',
  EDIT_HANDLE: 'editHandle',
  CURSOR_EDIT_HANDLE: 'cursorEditHandle'
};
exports.GUIDE_TYPE = GUIDE_TYPE;
var ELEMENT_TYPE = {
  FEATURE: 'feature',
  FILL: 'fill',
  SEGMENT: 'segment',
  EDIT_HANDLE: 'editHandle'
};
exports.ELEMENT_TYPE = ELEMENT_TYPE;
var EDIT_TYPE = {
  ADD_FEATURE: 'addFeature',
  ADD_POSITION: 'addPosition',
  REMOVE_POSITION: 'removePosition',
  MOVE_POSITION: 'movePosition',
  FINISH_MOVE_POSITION: 'finishMovePosition'
};
exports.EDIT_TYPE = EDIT_TYPE;
