// react-map-gl-draw/dist/edit-modes/base-mode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _constants = require("../constants");

var _utils = require("./utils");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var BaseMode =
/*#__PURE__*/
function () {
  function BaseMode() {
    var _this = this;

    _classCallCheck(this, BaseMode);

    _defineProperty(this, "_tentativeFeature", void 0);

    _defineProperty(this, "_editHandles", void 0);

    _defineProperty(this, "getTentativeFeature", function () {
      return _this._tentativeFeature;
    });

    _defineProperty(this, "getEditHandles", function () {
      return _this._editHandles;
    });

    _defineProperty(this, "setTentativeFeature", function (feature) {
      _this._tentativeFeature = feature;
    });

    _defineProperty(this, "getSelectedFeature", function (props, featureIndex) {
      var data = props.data,
          selectedIndexes = props.selectedIndexes;
      var featureCollection = data.getObject();
      var features = featureCollection && featureCollection.features;
      var selectedIndex = (0, _utils.isNumeric)(featureIndex) ? Number(featureIndex) : selectedIndexes && selectedIndexes[0];
      return features && features[selectedIndex];
    });

    this._tentativeFeature = null;
    this._editHandles = null;
  }

  _createClass(BaseMode, [{
    key: "handleClick",
    value: function handleClick(event, props) {}
  }, {
    key: "handlePointerMove",
    value: function handlePointerMove(event, props) {}
  }, {
    key: "handleStartDragging",
    value: function handleStartDragging(event, props) {}
  }, {
    key: "handleStopDragging",
    value: function handleStopDragging(event, props) {}
  }, {
    key: "getGuides",
    value: function getGuides(props) {}
  }, {
    key: "getEditHandlesFromFeature",
    value: function getEditHandlesFromFeature(feature, featureIndex) {
      var coordinates = (0, _utils.getFeatureCoordinates)(feature);

      if (!coordinates) {
        return null;
      }

      return coordinates.map(function (coord, i) {
        return {
          type: 'Feature',
          properties: {
            // TODO deprecate renderType
            renderType: feature.properties.renderType,
            guideType: _constants.GUIDE_TYPE.EDIT_HANDLE,
            featureIndex: featureIndex,
            positionIndexes: [i]
          },
          geometry: {
            type: _constants.GEOJSON_TYPE.POINT,
            coordinates: [coord]
          }
        };
      });
    }
  }]);

  return BaseMode;
}();

exports.default = BaseMode;
