// react-map-gl-draw/dist-es6/edit-modes/draw-polygon-mode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _v = _interopRequireDefault(require("uuid/v1"));

var _constants = require("../constants");

var _utils = require("./utils");

var _baseMode = _interopRequireDefault(require("./base-mode"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DrawPolygonMode =
/*#__PURE__*/
function (_BaseMode) {
  _inherits(DrawPolygonMode, _BaseMode);

  function DrawPolygonMode() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, DrawPolygonMode);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(DrawPolygonMode)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleClick", function (event, props) {
      var picked = event.picks && event.picks[0];
      var data = props.data; // update tentative feature

      var tentativeFeature = _this.getTentativeFeature(); // add position to tentativeFeature
      // if click the first editHandle, commit tentativeFeature to featureCollection


      if (tentativeFeature) {
        var pickedObject = picked && picked.object; // clicked an editHandle of a tentative feature

        if (pickedObject && pickedObject.index === 0) {
          _this.setTentativeFeature(null); // append point to the tail, close the polygon


          var coordinates = (0, _utils.getFeatureCoordinates)(tentativeFeature);

          if (!coordinates) {
            return;
          }

          coordinates.push(coordinates[0]);
          tentativeFeature = {
            type: 'Feature',
            properties: {
              // TODO deprecate id
              id: tentativeFeature.properties.id,
              renderType: _constants.RENDER_TYPE.POLYGON
            },
            geometry: {
              type: _constants.GEOJSON_TYPE.POLYGON,
              coordinates: [coordinates]
            }
          };
          var updatedData = data.addFeature(tentativeFeature).getObject();
          props.onEdit({
            editType: _constants.EDIT_TYPE.ADD_FEATURE,
            updatedData: updatedData,
            editContext: null
          });
        } else {
          // update tentativeFeature
          tentativeFeature = _objectSpread({}, tentativeFeature, {
            geometry: {
              type: _constants.GEOJSON_TYPE.LINE_STRING,
              coordinates: _toConsumableArray(tentativeFeature.geometry.coordinates).concat([event.mapCoords])
            }
          });

          _this.setTentativeFeature(tentativeFeature);
        }
      } else {
        // create a tentativeFeature
        tentativeFeature = {
          type: 'Feature',
          properties: {
            // TODO deprecate id
            id: (0, _v.default)(),
            renderType: _constants.RENDER_TYPE.POLYGON,
            guideType: _constants.GUIDE_TYPE.TENTATIVE
          },
          geometry: {
            type: _constants.GEOJSON_TYPE.POINT,
            coordinates: [event.mapCoords]
          }
        };

        _this.setTentativeFeature(tentativeFeature);
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getGuides", function (props) {
      var tentativeFeature = _this.getTentativeFeature();

      var coordinates = (0, _utils.getFeatureCoordinates)(tentativeFeature);

      if (!coordinates) {
        return null;
      }

      var event = props.lastPointerMoveEvent; // existing editHandles + cursorEditHandle

      var editHandles = _this.getEditHandlesFromFeature(tentativeFeature) || [];
      var cursorEditHandle = {
        type: 'Feature',
        properties: {
          guideType: _constants.GUIDE_TYPE.CURSOR_EDIT_HANDLE,
          // TODO remove renderType
          renderType: _constants.RENDER_TYPE.POLYGON,
          positionIndexes: [editHandles.length]
        },
        geometry: {
          type: _constants.GEOJSON_TYPE.POINT,
          coordinates: [event.mapCoords]
        }
      };
      editHandles.push(cursorEditHandle); // tentativeFeature

      tentativeFeature = _objectSpread({}, tentativeFeature, {
        geometry: {
          type: _constants.GEOJSON_TYPE.LINE_STRING,
          coordinates: _toConsumableArray(coordinates).concat([event.mapCoords])
        }
      });
      return {
        tentativeFeature: tentativeFeature,
        editHandles: editHandles
      };
    });

    return _this;
  }

  return DrawPolygonMode;
}(_baseMode.default);

exports.default = DrawPolygonMode;
