// react-map-gl-draw/dist-es6/edit-modes/draw-line-string-mode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _v = _interopRequireDefault(require("uuid/v1"));

var _constants = require("../constants");

var _baseMode = _interopRequireDefault(require("./base-mode"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DrawLineStringMode =
/*#__PURE__*/
function (_BaseMode) {
  _inherits(DrawLineStringMode, _BaseMode);

  function DrawLineStringMode() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, DrawLineStringMode);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(DrawLineStringMode)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "handleClick", function (event, props) {
      var data = props.data,
          selectedIndexes = props.selectedIndexes;

      var selectedFeature = _this.getSelectedFeature(props);

      var tentativeFeature = _this.getTentativeFeature(); // add position to a selectedFeature


      if (selectedFeature) {
        var selectedFeatureIndex = selectedIndexes[0];
        var positionIndexes = [selectedFeature.geometry.coordinates.length];
        var updatedData = data.addPosition(selectedFeatureIndex, positionIndexes, event.mapCoords).getObject();
        props.onEdit({
          editType: _constants.EDIT_TYPE.ADD_POSITION,
          updatedData: updatedData,
          editContext: [{
            featureIndex: selectedFeatureIndex,
            editHandleIndex: positionIndexes[0],
            screenCoords: event.screenCoords,
            mapCoords: event.mapCoords
          }]
        }); // commit tentativeFeature to featureCollection
      } else if (tentativeFeature) {
        _this.setTentativeFeature(null);

        var feature = {
          type: 'Feature',
          properties: {
            id: tentativeFeature.properties.id,
            // todo deprecate renderType
            renderType: _constants.RENDER_TYPE.LINE_STRING
          },
          geometry: {
            type: _constants.GEOJSON_TYPE.LINE_STRING,
            coordinates: [tentativeFeature.geometry.coordinates[0], event.mapCoords]
          }
        };

        var _updatedData = data.addFeature(feature).getObject();

        props.onEdit({
          editType: _constants.EDIT_TYPE.ADD_FEATURE,
          updatedData: _updatedData,
          editContext: null
        });
      } else {
        tentativeFeature = {
          type: 'Feature',
          properties: {
            // TODO deprecate id & renderType
            id: (0, _v.default)(),
            renderType: _constants.RENDER_TYPE.LINE_STRING,
            guideType: _constants.GUIDE_TYPE.TENTATIVE
          },
          geometry: {
            type: _constants.GEOJSON_TYPE.POINT,
            coordinates: [event.mapCoords]
          }
        };

        _this.setTentativeFeature(tentativeFeature);
      }
    });

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "getGuides", function (props) {
      var selectedFeature = _this.getSelectedFeature(props);

      var tentativeFeature = _this.getTentativeFeature();

      var feature = selectedFeature || tentativeFeature;
      var coordinates = (0, _utils.getFeatureCoordinates)(feature);

      if (!coordinates) {
        return null;
      }

      var event = props.lastPointerMoveEvent; // existing editHandles + cursorEditHandle

      var editHandles = _this.getEditHandlesFromFeature(feature) || [];
      var cursorEditHandle = {
        type: 'Feature',
        properties: {
          guideType: _constants.GUIDE_TYPE.CURSOR_EDIT_HANDLE,
          // TODO remove renderType
          renderType: _constants.RENDER_TYPE.LINE_STRING,
          positionIndexes: [editHandles.length]
        },
        geometry: {
          type: _constants.GEOJSON_TYPE.POINT,
          coordinates: [event.mapCoords]
        }
      };
      editHandles.push(cursorEditHandle); // tentativeFeature

      tentativeFeature = {
        type: 'Feature',
        properties: {
          // TODO deprecate id and renderType
          id: (0, _v.default)(),
          guideType: _constants.GUIDE_TYPE.TENTATIVE,
          renderType: _constants.RENDER_TYPE.LINE_STRING
        },
        geometry: {
          type: _constants.GEOJSON_TYPE.LINE_STRING,
          coordinates: [coordinates[coordinates.length - 1], event.mapCoords]
        }
      };
      return {
        tentativeFeature: tentativeFeature,
        editHandles: editHandles
      };
    });

    return _this;
  }

  return DrawLineStringMode;
}(_baseMode.default);

exports.default = DrawLineStringMode;
