// @nebula.gl/edit-modes/dist-es6/lib/composite-mode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CompositeMode = void 0;

var _geojsonEditMode = require("./geojson-edit-mode.js");

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance"); }

function _iterableToArray(iter) { if (Symbol.iterator in Object(iter) || Object.prototype.toString.call(iter) === "[object Arguments]") return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = new Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var CompositeMode =
/*#__PURE__*/
function (_BaseGeoJsonEditMode) {
  _inherits(CompositeMode, _BaseGeoJsonEditMode);

  function CompositeMode(modes) {
    var _this;

    _classCallCheck(this, CompositeMode);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(CompositeMode).call(this));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_modes", void 0);

    _this._modes = modes;
    return _this;
  }

  _createClass(CompositeMode, [{
    key: "_coalesce",
    value: function _coalesce(callback) {
      var resultEval = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
      var result;

      for (var i = 0; i < this._modes.length; i++) {
        result = callback(this._modes[i]);

        if (resultEval ? resultEval(result) : result) {
          break;
        }
      }

      return result;
    }
  }, {
    key: "handleClick",
    value: function handleClick(event, props) {
      this._coalesce(function (handler) {
        return handler.handleClick(event, props);
      });
    }
  }, {
    key: "handlePointerMove",
    value: function handlePointerMove(event, props) {
      return this._coalesce(function (handler) {
        return handler.handlePointerMove(event, props);
      });
    }
  }, {
    key: "handleStartDragging",
    value: function handleStartDragging(event, props) {
      return this._coalesce(function (handler) {
        return handler.handleStartDragging(event, props);
      });
    }
  }, {
    key: "handleStopDragging",
    value: function handleStopDragging(event, props) {
      return this._coalesce(function (handler) {
        return handler.handleStopDragging(event, props);
      });
    }
  }, {
    key: "handleDragging",
    value: function handleDragging(event, props) {
      return this._coalesce(function (handler) {
        return handler.handleDragging(event, props);
      });
    }
  }, {
    key: "getGuides",
    value: function getGuides(props) {
      // TODO: Combine the guides *BUT* make sure if none of the results have
      // changed to return the same object so that "guides !== this.state.guides"
      // in editable-geojson-layer works.
      var allGuides = [];
      var _iteratorNormalCompletion = true;
      var _didIteratorError = false;
      var _iteratorError = undefined;

      try {
        for (var _iterator = this._modes[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
          var mode = _step.value;
          allGuides.push.apply(allGuides, _toConsumableArray(mode.getGuides(props).features));
        }
      } catch (err) {
        _didIteratorError = true;
        _iteratorError = err;
      } finally {
        try {
          if (!_iteratorNormalCompletion && _iterator.return != null) {
            _iterator.return();
          }
        } finally {
          if (_didIteratorError) {
            throw _iteratorError;
          }
        }
      }

      return {
        type: 'FeatureCollection',
        features: allGuides
      };
    }
  }]);

  return CompositeMode;
}(_geojsonEditMode.BaseGeoJsonEditMode);

exports.CompositeMode = CompositeMode;
