// @nebula.gl/edit-modes/dist-es6/lib/measure-distance-mode.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MeasureDistanceMode = void 0;

var _distance = _interopRequireDefault(require("@turf/distance"));

var _memoizee = _interopRequireDefault(require("memoizee"));

var _geojsonEditMode = require("./geojson-edit-mode.js");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var DEFAULT_TOOLTIPS = [];

var MeasureDistanceMode =
/*#__PURE__*/
function (_BaseGeoJsonEditMode) {
  _inherits(MeasureDistanceMode, _BaseGeoJsonEditMode);

  function MeasureDistanceMode() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, MeasureDistanceMode);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(MeasureDistanceMode)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "startingPoint", null);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "endingPoint", null);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "endingPointLocked", false);

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "_getTooltips", (0, _memoizee.default)(function (modeConfig, startingPoint, endingPoint) {
      var tooltips = DEFAULT_TOOLTIPS;

      if (startingPoint && endingPoint) {
        var _ref = modeConfig || {},
            formatTooltip = _ref.formatTooltip,
            turfOptions = _ref.turfOptions,
            measurementCallback = _ref.measurementCallback;

        var units = turfOptions && turfOptions.units || 'kilometers';
        var distance = (0, _distance.default)(startingPoint, endingPoint, turfOptions);
        var text;

        if (formatTooltip) {
          text = formatTooltip(distance);
        }

        if (!formatTooltip) {
          // By default, round to 2 decimal places and append units
          text = "".concat(parseFloat(distance).toFixed(2), " ").concat(units);
        }

        if (measurementCallback) {
          measurementCallback(distance);
        }

        tooltips = [{
          position: endingPoint.geometry.coordinates,
          text: text
        }];
      }

      return tooltips;
    }));

    return _this;
  }

  _createClass(MeasureDistanceMode, [{
    key: "_setEndingPoint",
    value: function _setEndingPoint(mapCoords) {
      this.endingPoint = {
        type: 'Feature',
        properties: {
          guideType: 'editHandle',
          editHandleType: 'existing',
          featureIndex: -1,
          positionIndexes: []
        },
        geometry: {
          type: 'Point',
          coordinates: mapCoords
        }
      };
    }
  }, {
    key: "handleClick",
    value: function handleClick(event, props) {
      if (!this.startingPoint || this.endingPointLocked) {
        this.startingPoint = {
          type: 'Feature',
          properties: {
            guideType: 'editHandle',
            editHandleType: 'existing',
            featureIndex: -1,
            positionIndexes: []
          },
          geometry: {
            type: 'Point',
            coordinates: event.mapCoords
          }
        };
        this.endingPoint = null;
        this.endingPointLocked = false;
      } else if (this.startingPoint) {
        this._setEndingPoint(event.mapCoords);

        this.endingPointLocked = true;
      }
    } // Called when the pointer moved, regardless of whether the pointer is down, up, and whether something was picked

  }, {
    key: "handlePointerMove",
    value: function handlePointerMove(event, props) {
      if (this.startingPoint && !this.endingPointLocked) {
        this._setEndingPoint(event.mapCoords);
      }

      props.onUpdateCursor('cell');
    } // Called when the pointer went down on something rendered by this layer and the pointer started to move

  }, {
    key: "handleStartDragging",
    value: function handleStartDragging(event, props) {} // Called when the pointer went down on something rendered by this layer, the pointer moved, and now the pointer is up

  }, {
    key: "handleStopDragging",
    value: function handleStopDragging(event, props) {} // Return features that can be used as a guide for editing the data

  }, {
    key: "getGuides",
    value: function getGuides(props) {
      var guides = {
        type: 'FeatureCollection',
        features: []
      };
      var features = guides.features;

      if (this.startingPoint) {
        features.push(this.startingPoint);
      }

      if (this.endingPoint) {
        features.push(this.endingPoint);
      }

      if (this.startingPoint && this.endingPoint) {
        features.push({
          type: 'Feature',
          properties: {
            guideType: 'tentative'
          },
          geometry: {
            type: 'LineString',
            coordinates: [this.startingPoint.geometry.coordinates, this.endingPoint.geometry.coordinates]
          }
        });
      }

      return guides;
    }
  }, {
    key: "getTooltips",
    value: function getTooltips(props) {
      return this._getTooltips(props.modeConfig, this.startingPoint, this.endingPoint);
    }
  }]);

  return MeasureDistanceMode;
}(_geojsonEditMode.BaseGeoJsonEditMode);

exports.MeasureDistanceMode = MeasureDistanceMode;
